File: javascript/packages/baselode/src/data/drillholeSet.js
/*
 * Copyright (C) 2026 Darkmine Pty Ltd
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

/**
 * DrillholeSet — composition root for the drilling tables of a project.
 *
 * Holds a collar + survey table plus N named interval tables and exposes
 * the existing function-based API as methods.  No new logic; each method
 * delegates to the corresponding module (`validateDrillholeDb`,
 * `minimumCurvatureDesurvey`, etc.).  The trace is cached after the
 * first `desurvey()` call.
 *
 * Mirrors the Python `baselode.drill.DrillholeSet` class.
 */

import { HOLE_ID, EASTING, NORTHING, ELEVATION, DEPTH, FROM } from './datamodel.js';
import { validateDrillholeDb } from './validateDrillholeDb.js';
import {
  minimumCurvatureDesurvey,
  tangentialDesurvey,
  balancedTangentialDesurvey,
  midpointTangentialDesurvey,
} from './desurveyMethods.js';

const DESURVEY_METHODS = {
  minimum_curvature: minimumCurvatureDesurvey,
  tangential: tangentialDesurvey,
  balanced_tangential: balancedTangentialDesurvey,
  midpoint_tangential: midpointTangentialDesurvey,
};

// The JS desurvey helpers (kept compatible with their existing callers)
// read `collar.x/y/z` and `survey.from`. DrillholeSet stores rows in the
// canonical baselode shape (`easting/northing/elevation` and `depth`),
// so we map at the call boundary rather than mutating the stored tables.
function _collarToDesurveyShape(row, holeCol) {
  if (!row) return row;
  return {
    ...row,
    hole_id: row[holeCol] != null ? row[holeCol] : row.hole_id,
    x: row[EASTING] != null ? row[EASTING] : row.x,
    y: row[NORTHING] != null ? row[NORTHING] : row.y,
    z: row[ELEVATION] != null ? row[ELEVATION] : row.z,
  };
}

function _surveyToDesurveyShape(row, holeCol) {
  if (!row) return row;
  return {
    ...row,
    hole_id: row[holeCol] != null ? row[holeCol] : row.hole_id,
    [FROM]: row[DEPTH] != null ? row[DEPTH] : row[FROM],
  };
}

export class DrillholeSet {
  /**
   * @param {Array<Object>} collar - Collar rows
   * @param {Array<Object>} survey - Survey rows
   * @param {Object} [options]
   * @param {string} [options.crs]
   * @param {string} [options.project]
   * @param {string} [options.holeCol=HOLE_ID]
   */
  constructor(collar, survey, { crs = null, project = null, holeCol = HOLE_ID } = {}) {
    this.collar = collar || [];
    this.survey = survey || [];
    this.crs = crs;
    this.project = project;
    this.holeCol = holeCol;
    this.tables = {};
    this.tableKinds = {};
    this._traces = null;
    this._desurveyArgs = null;
  }

  /**
   * Register an interval table.  Chainable.
   */
  addTable(name, rows, kind = 'assay') {
    if (typeof name !== 'string' || !name) {
      throw new Error('Table name must be a non-empty string');
    }
    if (name === 'collars' || name === 'traces') {
      throw new Error(`Table name '${name}' is reserved for built-in OMF elements`);
    }
    this.tables[name] = rows || [];
    this.tableKinds[name] = kind;
    return this;
  }

  /**
   * Distinct hole IDs from the collar table.
   */
  get holes() {
    const seen = new Set();
    for (const row of this.collar) {
      const value = row && row[this.holeCol];
      if (value == null) continue;
      seen.add(value);
    }
    return [...seen];
  }

  /**
   * Cached desurveyed trace; runs desurvey() on first access.
   */
  get traces() {
    if (this._traces === null) return this.desurvey();
    return this._traces;
  }

  /**
   * Desurvey the holes.  Cached when called with identical args.
   *
   * @param {Object} [options]
   * @param {string} [options.method='minimum_curvature']
   * @param {number} [options.step=1.0]
   * @param {boolean} [options.force=false]
   */
  desurvey({ method = 'minimum_curvature', step = 1.0, force = false } = {}) {
    const args = `${method}:${step}`;
    if (this._traces !== null && this._desurveyArgs === args && !force) {
      return this._traces;
    }
    const methodFn = DESURVEY_METHODS[method];
    if (!methodFn) {
      const available = Object.keys(DESURVEY_METHODS).sort().join(', ');
      throw new Error(`Unknown desurvey method '${method}'. Available: ${available}`);
    }
    const collarShaped = this.collar.map((row) => _collarToDesurveyShape(row, this.holeCol));
    const surveyShaped = this.survey.map((row) => _surveyToDesurveyShape(row, this.holeCol));
    this._traces = methodFn(collarShaped, surveyShaped, { step });
    this._desurveyArgs = args;
    return this._traces;
  }

  /**
   * Run validateDrillholeDb over the set.  Extra options are forwarded.
   */
  validate(options = {}) {
    return validateDrillholeDb(
      {
        collar: this.collar,
        survey: this.survey,
        intervalTables: Object.keys(this.tables).length ? this.tables : null,
      },
      { holeCol: this.holeCol, ...options },
    );
  }

  /**
   * Array-style table accessor.
   */
  get(tableName) {
    return this.tables[tableName];
  }

  has(tableName) {
    return Object.prototype.hasOwnProperty.call(this.tables, tableName);
  }

  toString() {
    return (
      `<DrillholeSet holes=${this.holes.length} `
      + `surveyRows=${this.survey.length} `
      + `tables=[${Object.keys(this.tables).join(', ')}]>`
    );
  }
}

File: python/src/baselode/drill/drillhole_set.py
# Copyright (C) 2026 Darkmine Pty Ltd
# SPDX-License-Identifier: GPL-3.0-or-later

"""DrillholeSet — composition root for the drilling tables of a project.

Bundles a collar + survey table plus an arbitrary number of named interval
tables (assay, geology, structural, geotech, ...) and exposes the
existing function-based API as methods.  No new algorithmic logic: every
method delegates to ``baselode.drill.validate`` / ``desurvey`` /
``composite`` / ``omf``.  The bare-DataFrame functions remain importable
and unchanged.

State is restricted to derived data: the desurveyed trace is cached on
the instance so downstream methods (``to_omf``, future block-model fits)
don't repeatedly re-run desurvey.

Mirrors PyGSLIB's ``Drillhole(collar, survey)`` + ``addtable(...)``
ergonomics, which is the established mental model in the
geology/geometallurgy community.
"""

from baselode.datamodel import HOLE_ID


def _desurvey_methods():
    """Lazy registry of method-name → desurvey function.

    Lazy so the import doesn't trigger while ``baselode.drill`` itself
    is still loading (this module is re-exported from
    ``baselode.drill.__init__``).
    """
    import baselode.drill.desurvey as desurvey
    return {
        "minimum_curvature": desurvey.minimum_curvature_desurvey,
        "tangential": desurvey.tangential_desurvey,
        "balanced_tangential": desurvey.balanced_tangential_desurvey,
    }


class DrillholeSet:
    """A collection of related drilling tables for one project / dataset.

    Parameters
    ----------
    collar : pd.DataFrame
        Collar table — must carry ``hole_id``, ideally with the projected
        ``easting`` / ``northing`` / ``elevation`` and any metadata
        columns (max_depth, hole_type, project_id, ...).
    survey : pd.DataFrame
        Directional survey table with ``hole_id``, ``depth``,
        ``azimuth``, ``dip``.
    crs : str or None
        Optional EPSG code / proj string travelling with the set.
    project : str or None
        Optional project identifier.

    Attributes
    ----------
    collar, survey : pd.DataFrame
        The source tables.
    tables : dict[str, pd.DataFrame]
        Interval tables registered via :meth:`add_table`.
    table_kinds : dict[str, str]
        Per-table kind tag (e.g. ``"assay"``, ``"litho"``, ``"structural"``).
    crs, project : str or None
        Metadata passed at construction.
    holes : list
        Distinct hole IDs from the collar table (property).
    traces : pd.DataFrame or None
        Cached desurvey output.  ``None`` until :meth:`desurvey` runs.
    """

    def __init__(self, collar, survey, crs=None, project=None, hole_col=HOLE_ID):
        self.collar = collar
        self.survey = survey
        self.crs = crs
        self.project = project
        self.hole_col = hole_col
        self.tables = {}
        self.table_kinds = {}
        self._traces = None
        self._desurvey_args = None

    def add_table(self, name, df, kind="assay"):
        """Register an interval table.  Chainable.

        Parameters
        ----------
        name : str
            Unique key; existing entries are overwritten.  ``"collars"``
            and ``"traces"`` are reserved for the built-in OMF elements
            in :meth:`to_omf` and raise :class:`ValueError`.
        df : pd.DataFrame
        kind : str
            Free-form tag describing the table (default ``"assay"``).
            Stored alongside the table for downstream consumers; not
            consulted by the methods on this class.
        """
        if not isinstance(name, str) or not name:
            raise ValueError("Table name must be a non-empty string")
        if name in ("collars", "traces"):
            raise ValueError(
                f"Table name '{name}' is reserved for built-in OMF elements"
            )
        self.tables[name] = df
        self.table_kinds[name] = kind
        return self

    @property
    def holes(self):
        """Distinct hole IDs from the collar table, NaN dropped."""
        if self.hole_col not in self.collar.columns:
            return []
        return self.collar[self.hole_col].dropna().unique().tolist()

    @property
    def traces(self):
        """Cached desurveyed trace; computes on first access using defaults."""
        if self._traces is None:
            return self.desurvey()
        return self._traces

    def desurvey(self, method="minimum_curvature", step=1.0, force=False):
        """Desurvey the holes (cached after first call).

        Parameters
        ----------
        method : str
            One of ``"minimum_curvature"``, ``"tangential"``,
            ``"balanced_tangential"``.
        step : float
            Output sample spacing (metres).
        force : bool
            Re-run even when a cached result with matching args exists.

        Returns
        -------
        pd.DataFrame
            Trace table with ``hole_id``, ``md``, ``easting``,
            ``northing``, ``elevation``, ``azimuth``, ``dip``.
        """
        args = (method, step)
        if self._traces is not None and self._desurvey_args == args and not force:
            return self._traces
        methods = _desurvey_methods()
        method_fn = methods.get(method)
        if method_fn is None:
            available = ", ".join(sorted(methods))
            raise ValueError(f"Unknown desurvey method '{method}'. Available: {available}")
        self._traces = method_fn(self.collar, self.survey, step=step)
        self._desurvey_args = args
        return self._traces

    def validate(self, **kwargs):
        """Run :func:`baselode.drill.validate.validate_drillhole_db` over the set.

        All keyword arguments are forwarded to the underlying function
        (e.g. ``allow_full_circle=True``).
        """
        import baselode.drill.validate as validate_module
        return validate_module.validate_drillhole_db(
            self.collar,
            self.survey,
            interval_tables=self.tables or None,
            **kwargs,
        )

    def composite(self, table_name, value_col, length=1.0, method="average", from_col="from", to_col="to"):
        """Composite an interval table to a fixed downhole length.

        Thin wrapper around
        :func:`baselode.drill.composite.composite_intervals`.
        """
        if table_name not in self.tables:
            raise KeyError(
                f"No table '{table_name}'; registered: {list(self.tables)}"
            )
        import baselode.drill.composite as composite_module
        return composite_module.composite_intervals(
            self.tables[table_name],
            value_col,
            from_col=from_col,
            to_col=to_col,
            length=length,
            method=method,
        )

    def to_omf(
        self,
        path,
        include=None,
        value_cols=None,
        name=None,
        author="baselode",
        description="",
    ):
        """Serialise the set to an OMF v1 file.

        Parameters
        ----------
        path : str or pathlib.Path
        include : iterable of str, optional
            Element keys to include — any of ``"collars"``, ``"traces"``,
            or a name from :attr:`tables`.  Defaults to everything that
            has been registered.
        value_cols : dict[str, list[str]], optional
            Per-table value columns to attach as OMF segment data, e.g.
            ``{"assay": ["au_ppm", "cu_pct"]}``.  Tables absent from this
            mapping get no value columns (just per-segment ``hole_id``).
        name : str, optional
            OMF project name; defaults to :attr:`project` or
            ``"baselode"``.
        author, description : str
            OMF project metadata.

        Returns
        -------
        pathlib.Path
            The path written.
        """
        if include is None:
            include = ["collars", "traces"] + list(self.tables)
        else:
            include = list(include)
        value_cols = dict(value_cols or {})

        import baselode.drill.omf as omf_module
        elements = []
        if "collars" in include:
            elements.append(omf_module.collars_to_omf_points(self.collar))
        if "traces" in include:
            elements.append(omf_module.traces_to_omf_lines(self.traces))
        for table_name in include:
            if table_name in ("collars", "traces"):
                continue
            if table_name not in self.tables:
                raise KeyError(
                    f"No table '{table_name}'; registered: {list(self.tables)}"
                )
            elements.append(omf_module.intervals_to_omf_lines(
                self.tables[table_name],
                self.traces,
                name=table_name,
                value_cols=value_cols.get(table_name),
            ))

        from pathlib import Path
        path = Path(path)
        omf_module.write_omf(
            elements,
            path,
            name=name or self.project or "baselode",
            author=author,
            description=description,
        )
        return path

    def __getitem__(self, table_name):
        return self.tables[table_name]

    def __contains__(self, table_name):
        return table_name in self.tables

    def __repr__(self):
        return (
            f"<DrillholeSet holes={len(self.holes)} "
            f"survey_rows={len(self.survey)} "
            f"tables={list(self.tables)}>"
        )
